### PR TITLE
Removing date segment placeholder default width

### PR DESCRIPTION
**Pull request by ktabors**
Closes 

## ✅ Pull Request Checklist:

- [ ] Included link to corresponding [React Spectrum GitHub Issue](https://github.com/adobe/react-spectrum/issues).
- [ ] Added/updated unit tests and storybook for this change (for new code or code which already has tests).
- [ ] Filled out test instructions.
- [ ] Updated documentation (if it already exists for this component).
- [ ] Looked at the Accessibility Practices for this feature - [Aria Practices](https://www.w3.org/WAI/ARIA/apg/)

### CODE REVIEW

File: packages/@react-spectrum/datepicker/src/DatePickerSegment.tsx [hunk 1]
```diff
@@ -60,13 +60,9 @@ function EditableSegment({segment, state}: DatePickerSegmentProps) {
...
       })}
-      style={{
-        ...segmentProps.style,
-        minWidth: segment.maxValue != null ? String(segment.maxValue).length + 'ch' : null
-      }}
+      style={segmentProps.style}
       data-testid={segment.type}>
-      <span aria-hidden="true" className={classNames(styles, 'react-spectrum-DatePicker-placeholder')}>{segment.placeholder}</span>
-      {segment.isPlaceholder ? '' : segment.text}
+      {segment.isPlaceholder ? <span aria-hidden="true" className={classNames(styles, 'react-spectrum-DatePicker-placeholder')}>{segment.placeholder}</span> : segment.text}
```
Comment (devongovett) [inline packages/@react-spectrum/datepicker/src/DatePickerSegment.tsx:65]: I'm not seeing this span in the DOM at all, not sure why. It looks like the placeholder text (e.g. mm) is directly inside the div. 🤔

File: packages/@react-spectrum/datepicker/src/styles.css [hunk 2]
```diff
@@ -111,22 +111,13 @@
...
-}
-
 .react-spectrum-DatePicker-cell.is-placeholder {
   color: var(--spectrum-gray-600);
 
   .react-spectrum-DatePicker-placeholder {
-    visibility: visible;
-    height: auto;
+    text-align: center;
+    font-style: italic;
```
Comment (devongovett) [inline packages/@react-spectrum/datepicker/src/styles.css:119]: This element isn't working at all right now so I'm not seeing the italics. But I guess we should get rid of them anyway?

**Comment by devongovett** [comment]
One other problem appears to be that when you type into the field, the entire field gets wider or shorter...

File: packages/@react-spectrum/datepicker/src/DatePickerSegment.tsx (hunk 1 above)
Comment (ktabors) [reply packages/@react-spectrum/datepicker/src/DatePickerSegment.tsx:65]: Which build are you testing? When I look at bc8e7a2a8d6b89da4f821cefe50488b494433f35 I see the following.
image

Comment (devongovett) [reply packages/@react-spectrum/datepicker/src/DatePickerSegment.tsx:65]: hmm weird, it is working now... not sure what happened before.

Should we get rid of the italics though while we're at it? @dannify seemed to think so

File: packages/@react-spectrum/datepicker/src/styles.css (hunk 2 above)
Comment (ktabors) [reply packages/@react-spectrum/datepicker/src/styles.css:119]: Ditto the above for not seeing italics.
We are purging italics, but do we need that visual indicator for placeholder?

File: packages/@react-stately/datepicker/src/useDatePickerState.ts [hunk 3]
```diff
@@ -62,7 +62,9 @@ export interface DatePickerState extends OverlayTriggerState, FormValidationStat
   /** Whether the date picker is invalid, based on the `isInvalid`, `minValue`, and `maxValue` props. */
   isInvalid: boolean,
   /** Formats the selected value using the given options. */
-  formatValue(locale: string, fieldOptions: FieldOptions): string
+  formatValue(locale: string, fieldOptions: FieldOptions): string,
+  /** Gets a formatter based on state's props. */
+  getDateFormatter(ocale: string, formatOptions: FormatterOptions): DateFormatter
```
Comment (snowystinger) [inline packages/@react-stately/datepicker/src/useDatePickerState.ts:67]:
```suggestion
  getDateFormatter(locale: string, formatOptions: FormatterOptions): DateFormatter
```

File: packages/@react-spectrum/datepicker/src/utils.ts [hunk 4]
```diff
@@ -77,3 +77,16 @@ export function useFocusManagerRef(ref: FocusableRef<HTMLElement>) {
...
+export function useDateCharacterWidth(state) {
+  let locale = useLocale()?.locale;
+  let currentDate = new Date();
+  let formatedDate = state.getDateFormatter(locale, {shouldForceLeadingZeros: true}).format(currentDate, locale);
+  let totalCharacters =  formatedDate.length;
+
+  // The max of two is for times with only hours.
+  // As the length of a date grows we need to proportionally increase the width.
+  // We use the characts with 'ch' and months and time dashes are wider and this
+  // accomates for that.
```
Comment (snowystinger) [inline packages/@react-spectrum/datepicker/src/utils.ts:90]:
```suggestion
  // accommodates for that.
```

File: packages/@react-spectrum/datepicker/src/utils.ts [hunk 5]
```diff
@@ -77,3 +77,16 @@ export function useFocusManagerRef(ref: FocusableRef<HTMLElement>) {
...
+
+export function useDateCharacterWidth(state) {
+  let locale = useLocale()?.locale;
+  let currentDate = new Date();
+  let formatedDate = state.getDateFormatter(locale, {shouldForceLeadingZeros: true}).format(currentDate, locale);
+  let totalCharacters =  formatedDate.length;
+
+  // The max of two is for times with only hours.
+  // As the length of a date grows we need to proportionally increase the width.
+  // We use the characts with 'ch' and months and time dashes are wider and this
```
Comment (snowystinger) [inline packages/@react-spectrum/datepicker/src/utils.ts:89]:
```suggestion
  // We use the characters width 'ch' and months and time dashes are wider and this
```
?? this sentence didn't make much sense

I assume this is trying to explain why we do a divide by 5? i don't really follow why that is. Maybe some examples would help?

File: packages/@react-spectrum/datepicker/src/TimeField.tsx [hunk 6]
```diff
@@ -52,6 +52,8 @@ function TimeField<T extends TimeValue>(props: SpectrumTimeFieldProps<T>, ref: F
 
   let validationState = state.validationState || (isInvalid ? 'invalid' : null);
 
+  let characterCount = useDateCharacterWidth(state) + 'ch';
```
Comment (snowystinger) [inline packages/@react-spectrum/datepicker/src/TimeField.tsx:55]: this isn't actually a character count though is it? nor a character width?

File: packages/@react-spectrum/datepicker/src/utils.ts [hunk 7]
```diff
@@ -77,3 +77,16 @@ export function useFocusManagerRef(ref: FocusableRef<HTMLElement>) {
   }));
   return domRef;
 }
+
+export function useDateCharacterWidth(state) {
```
Comment (snowystinger) [inline packages/@react-spectrum/datepicker/src/utils.ts:81]: is it actually a character width? or is it the width of the characters + some fudging number? is this actually `useDateSegmentWidth`?

File: packages/@react-spectrum/datepicker/src/DateRangePicker.tsx [hunk 8]
```diff
@@ -112,6 +112,9 @@ function DateRangePicker<T extends DateValue>(props: SpectrumDateRangePickerProp
   let visibleMonths = useVisibleMonths(maxVisibleMonths);
   let validationState = state.validationState || (isInvalid ? 'invalid' : null);
 
+  // Multiplying by two for the two dates, adding one character for the dash, and then the dash left and right padding
```
Comment (snowystinger) [inline packages/@react-spectrum/datepicker/src/DateRangePicker.tsx:115]:
```suggestion
  // Multiplying by two for the two dates, adding one character for the dash, and then the padding around the dash
```

File: packages/@react-spectrum/datepicker/src/utils.ts (hunk 5 above)
Comment (LFDanLu) [reply packages/@react-spectrum/datepicker/src/utils.ts:89]: I don't quite follow the calculation either, I think I've formatted as wide of a date as possible but the min width still gives some extra padding here
image

The character count is 24 and the calculated character width with the additional accommodation factor sets the width to 28ch which gives it that extra purple padding. Using 24ch makes it pretty flush with the inputted date
image
which seems better?

**Review by LFDanLu — COMMENTED** [review]
image
Feels like a lot of whitespace on the end but I don't think there is much we can do about that without making the field width shift which feels worse haha.

I also found that Android now crashes when trying to change AM/PM in the time field. Steps to reproduce:
1. Go to https://reactspectrum.blob.core.windows.net/reactspectrum/1a948f002e8a7279c153110cbe0b9d4f4e8f5820/storybook/index.html?path=/story/date-and-time-timefield--default&providerSwitcher-express=false
2. Fill out the field with any time
3. Focus the AM/PM field and hit backspace to clear it. Focus should move on to the previous segment
4. tap the AM/PM segment and try to change it to PM by pressing P. Note it doesn't change. 
5. Press P again and note it crashes

File: packages/@react-spectrum/datepicker/src/utils.ts (hunk 5 above)
Comment (devongovett) [reply packages/@react-spectrum/datepicker/src/utils.ts:89]: I think the problem is when you change it to a placeholder. "mm" is much wider than two numbers, so it would push beyond the min-width causing the field to grow.

File: packages/@react-spectrum/datepicker/src/TimeField.tsx (hunk 6 above)
Comment (ktabors) [reply packages/@react-spectrum/datepicker/src/TimeField.tsx:55]: Both... It's the character count being turned into a width. And on Picker it's a `calc()` which makes it more width.

File: packages/@react-spectrum/datepicker/src/utils.ts (hunk 5 above)
Comment (LFDanLu) [reply packages/@react-spectrum/datepicker/src/utils.ts:89]: still seems to have a bit of extra space by ~1 or 2ch:
image
I'm fine with it being a bit off/not exactly precise, just wanna understand the underlying calculation here to see if there is any way we can make it more precise, the white space after the field is already pretty large compared to before (of which people already complained about haha)

File: packages/@react-spectrum/datepicker/src/utils.ts (hunk 7 above)
Comment (ktabors) [reply packages/@react-spectrum/datepicker/src/utils.ts:81]: It's more like `useApproximateFormattedDateWidth`, but that is too long. I'll update it along with `characterCount` to something better.

File: packages/@react-spectrum/datepicker/src/utils.ts (hunk 5 above)
Comment (ktabors) [reply packages/@react-spectrum/datepicker/src/utils.ts:89]: TLDR: This works remarkably well as a guess/approximation and is because of the width of month and dash characters in placeholder values.

It does widen a few cases. I tried a bunch of things with ch and mm units, including a granularity-based switch statement. This ratio works well, and it is remarkably close. Since we're guessing and using today's date, not the placeholder or the actual date value, I'm not sure if getting more precise would be better.

When Devon and I were looking at it and talking about it yesterday, it ended up being that we need approximately two characters for any data value less than 10 characters. For example, we have a problem with TimeField width shifting for granularity hour if we don't give it four characters of space. Like Devon stated above, for dates it's the months and for time's it's the dashes that create the need for extra characters. As we grow past ten characters, we're adding time which is dashes so we need more characters. 

Also, this seems to work with locales, but we don't have a collection of rigorous tests with all the possible font, language, and calendar combinations. Getting more precise for English might cause problems elsewhere, while being more generous has worked so for with the locale testing that has been done.

Yes, it looks "bad" when you have a DateRange with granularity seconds with two values, where it only fills about half the field, but when you switch it to placeholder it fills it back up. We're shifting from all the white space being in the segments to the white space being after the date values. The field widths didn't change much. Looking at chromatic most of the field widths didn't grow in width and it was only by a little for those that did. It was about equivalent width grown to the purple space in that image Daniel shared.

**Comment by ktabors** [comment]
> I also found that Android now crashes when trying to change AM/PM in the time field. Steps to reproduce:

I was able to reproduce it on an another PR build. Follow the same steps, but when it doesn't crash press backspace and then it crashes. We should enter a bug for this.

File: packages/@react-spectrum/datepicker/src/utils.ts (hunk 7 above)
Comment (snowystinger) [reply packages/@react-spectrum/datepicker/src/utils.ts]: I'm fine with the hook name being long, it's private and i'd prefer the name be accurate

File: packages/@react-spectrum/datepicker/src/utils.ts (hunk 5 above)
Comment (snowystinger) [reply packages/@react-spectrum/datepicker/src/utils.ts]: Instead of guesstimating, could we make use of useLayoutEffect to render the placeholder, measure, then replace the placeholder with the real value before paint actually happens? Then we should be exact.

We could also render it position absolute and visibility hidden behind the real one, that way we don't need to make a change to state and it'd pick up all the correct CSS inheritance

**Review by snowystinger — COMMENTED** [review]
I assume we're ok with the minWidth on all these fields changing in response to the time? Especially for ones with shorter durations, like fields containing seconds. Or even just moving from AM -> PM I think was noted?

Sounds like it might cause problems for chromatic unless we mock the time for those?

**Review by akyllus — APPROVED** [review]

File: packages/@react-spectrum/datepicker/src/utils.ts (hunk 5 above)
Comment (ktabors) [reply packages/@react-spectrum/datepicker/src/utils.ts]: That was my first approach. I recall there being edge cases, it broke a lot of tests even with aria-hidden, and we still have the DateRangePicker problem of needing to set the value higher up than we have the value, placeholder, and segments.

**Review by LFDanLu — APPROVED** [review]
Approving for testing, filed https://github.com/orgs/adobe/projects/19/views/4?pane=issue&itemId=70985842 for the Android crash since it is pre-existing.

**Comment by yihuiliao** [comment]
> filed https://github.com/orgs/adobe/projects/19/views/4?pane=issue&itemId=70985842 for the Android crash

Seems like this bug is basically https://github.com/adobe/react-spectrum/issues/6183 except that instead of crashing, it concatenates AM and PM (so that it looks like AMPM)

**Review by yihuiliao — APPROVED** [review]
Approving for testing
